components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/SetAccountAssociationToLocalUserImpl.java

```diff
@@ -0,0 +1,84 @@
+import org.wso2.carbon.identity.user.profile.mgt.UserProfileException;
+import org.wso2.carbon.identity.user.profile.mgt.dao.UserProfileMgtDAO;
+
+public class SetAccountAssociationToLocalUserImpl implements SetAccountAssociationToLocalUser {
+
+    private static final Log log = LogFactory.getLog(SetAccountAssociationToLocalUserImpl.class);
+
+    @Override
+    public void doAssociationWithLocalUser(JsAuthenticatedUser federatedUser, String username, String tenantDomain,
+                                           String userStoreDomainName) {
+
+        if (federatedUser != null) {
+            if (!federatedUser.getWrapped().isFederatedUser()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User " + federatedUser.getWrapped().getUserName() + " is not a federated user.");
+                }
+            }
+            String externalSubject = federatedUser.getWrapped().getAuthenticatedSubjectIdentifier();
+            String externalIdpName = federatedUser.getWrapped().getFederatedIdPName();
+            if (externalSubject != null && externalIdpName != null) {
+                try {
+                    associateID(externalIdpName, externalSubject, username, tenantDomain, userStoreDomainName);
+                } catch (UserProfileException e) {
+                    log.error("Error while associate the local user to : " + externalSubject, e);
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(" Authenticated user or External IDP may be null " + " Authenticated User: " +
+                            externalSubject + " and the External IDP name: " + externalIdpName);
+                }
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug(" Federated user is null ");
+            }
+        }
+    }
+
+    /**
+     * @param idpID               external IDP name
+     * @param associatedID        external authenticated user
+     * @param username            local user name
+     * @param tenantDomain        tenant domain
+     * @param userStoreDomainName user store domain name
+     * @throws UserProfileException Error occurred while retrieving user profile
+     */
+    private void associateID(String idpID, String associatedID, String username, String tenantDomain,
+                             String userStoreDomainName) throws UserProfileException {
+
+        int tenantID = IdentityTenantUtil.getTenantId(tenantDomain);
+        try {
+            UserProfileMgtDAO userProfileMgtDAO = UserProfileMgtDAO.getInstance();
+            userProfileMgtDAO.createAssociation(tenantID, userStoreDomainName, username, idpID, associatedID);
+        } catch (UserProfileException e) {
+            String msg = "Error while creating association for user: " + username + " with federated IdP: " + "" + idpID;
+            log.error(msg, e);
+            throw new UserProfileException(msg, e);
```

Comment:
We are logging the error then in the doAssociationWithLocalUser, we are logging it again. We need to log the error in one place. So no need to catch the exception here. We can remove the try catch block